crherman7/rechunk
# feat: add cache versioning to Metro Bundler and refactor Babel plugin

## Scope



[closes TICKET-62](https://github.com/crherman7/rechunk/issues/62)

## Implementation



This PR introduces cache versioning to the Metro Bundler configuration and refactors the Babel plugin to improve maintainability, reduce redundancy, and enhance developer experience. Key updates include:

Metro Bundler Enhancements:
- Integrated `getCacheVersion` to generate a deterministic cache version based on:
  - Last modified time of `rechunk.json`.
  - The `RECHUNK_ENVIRONMENT` environment variable.
  - The running status of the Rechunk development server.
- Added `@crherman7/rechunk-babel-plugin` to `metro.config.js`.
- Updated Metro configuration to include `config.cacheVersion` for better cache consistency.

Babel Plugin Refactor:
- Moved `getCacheVersion` and `isRechunkDevServerRunning` logic to a new `lib/cache.ts` module.
- Simplified the `src/index.ts` by removing duplicated logic.
- Exported cache-related utilities via `lib/index.ts`.

General Improvements:
- Simplified npm scripts in `package.json` by removing redundant `-c` flags.
- Added detailed TSDoc comments to all exported functions and constants, improving code clarity.
- Improved developer experience with more concise and maintainable configurations.

These changes aim to enhance build consistency, streamline development workflows, and improve the clarity and maintainability of the codebase. 

**NOTE** Merge after https://github.com/crherman7/rechunk/pull/61

## Screenshots

<details>
  <summary>ios</summary>

  

![Simulator Screenshot - iPhone 15 - 2024-11-15 at 21 22 49](https://github.com/user-attachments/assets/bc7a9c13-96b0-4e3a-8469-39125da4c6e4)
</details>

<details>
  <summary>android</summary>
  
  

![Screenshot_1731723824](https://github.com/user-attachments/assets/31b5d4c5-f2df-4ea9-8a43-4cb8d0144d2d)
</details>

## How to Test



- Run the example app with `pnpm start:offline`
- Verify app runs and components work with hot reload w/ making changes
- Run the example app with `pnpm start:dev`
- Verify error from babel
- Run dev-server and the example app with `pnpm start:dev`
- Verify app runs

## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!

> 🟡 Why was the default value removed?

E.g: Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | RED |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |
